### PR TITLE
[ROCM] Support Multi-GPU offline tuning in TunableOp

### PR DESCRIPTION
**Pull request by naromero77amd**
This PR enhances offline tuning to support multi-GPUs. 

High-level description of algorithm:
- Duplicate GEMMs are first eliminated
- GEMMs are distributed to multi-GPUs for tuning
- Results are gathered into a file with `_full` in the filename

Also adding support for GemmAndBias and ScaledGemm

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Comment by naromero77amd** [comment]
Confirming that new test ran successfully in linux-focal-rocm6.2-py3.10/test (default, 2, 6, linux.rocm.gpu.2)
```
test/test_linalg.py::TestLinalgCUDA::test_matmul_check_entries_tunableop_cuda_float16, test/test_linalg.py::TestLinalgCUDA::test_matmul_offline_mgpu_tunableop_cuda_float32, 
```

**Comment by jpvillam-amd** [comment]
Tested this on 13k GEMMs on a 8 GPU machine and it worked great

**Comment by jfactory07** [comment]
Tested on 2 GPUs, it works well.

File: torch/cuda/tunable.py [hunk 1]
```diff
@@ -377,10 +371,37 @@ def _process_single_offline_gemm(untuned_gemm_line: str, gpu_id: int) -> None:
...
+
+    # Initialize dtype to make linter happy
+    dtype = None
+    dtypeA = None
+    dtypeB = None
+    dtypeC = None
+
+    if underscore_count == 2:
+        [op_sig, data_type, layout] = untuned_gemm[0].split("_")
+        transA = True if layout[0] == "T" else False
```
Comment (hongxiayang) [inline torch/cuda/tunable.py:388]:
```suggestion
        transA = layout[0] == "T"
```

File: torch/cuda/tunable.py [hunk 2]
```diff
@@ -377,10 +371,37 @@ def _process_single_offline_gemm(untuned_gemm_line: str, gpu_id: int) -> None:
...
+        transB = True if layout[1] == "T" else False
+        dtype = dtype_dict.get(data_type)
+    else:  # ScaledGEMM
+        untuned_gemm_temp = untuned_gemm[0].split("_")
+        op_sig = untuned_gemm_temp[0]
+        data_typeA = untuned_gemm_temp[1] + "_" + untuned_gemm_temp[2]
+        data_typeB = untuned_gemm_temp[3] + "_" + untuned_gemm_temp[4]
+        data_typeC = untuned_gemm_temp[5] + "_" + untuned_gemm_temp[6]
+        transA = True if untuned_gemm_temp[7][0] == "T" else False
```
Comment (hongxiayang) [inline torch/cuda/tunable.py:397]:
```suggestion
        transA = untuned_gemm_temp[7][0] == "T"
```

File: torch/cuda/tunable.py [hunk 3]
```diff
@@ -408,6 +428,42 @@ def _process_single_offline_gemm(untuned_gemm_line: str, gpu_id: int) -> None:
         matA = matA.transpose(1, 2) if transB else matA
         matB = matB.transpose(1, 2) if transA else matB
         torch.bmm(matA, matB)
+    elif op_sig == "ScaledGemmTunableOp":
+        fillA = 0.25
```
Comment (hongxiayang) [inline torch/cuda/tunable.py:432]: Can you explain the logic behind each of the constants?

**Comment by naromero77amd** [comment]
I am confirming that the test runs in the distributed configuration, e.g. in `trunk / linux-focal-rocm6.2-py3.10 / test (distributed, 1, 1, linux.rocm.gpu) (push)` 

```
Executing ['/opt/conda/envs/py_3.10/bin/python', '-bb', 'test_linalg.py', '--shard-id=1', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-k', 'test_matmul_offline_mgpu_tunable', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests'] ... [2024-11-23 23:07:27.500452]

test_linalg 1/1 was successful, full logs can be found in artifacts with path test/test-reports/test_linalg_1.1_f7422312d35486cb_.log
Running 1 items in this shard: test/test_linalg.py::TestLinalgCUDA::test_matmul_offline_mgpu_tunableop_cuda_float32
```
but also in one of the shards of `periodic / linux-focal-rocm6.2-py3.10 / test (distributed, 1, 3, linux.rocm.gpu, module:rocm, oncall:distributed)`

**Review by jeffdaily — APPROVED** [review]

**Review by hongxiayang — APPROVED** [review]

**Pull request closed without merging**